fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,139 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    fetch('/api/todos')
+      .then(res => {
+        if (!res.ok) {
+          throw new Error('Something went wrong.');
+        } else {
+          return res.json();
+        }
+      })
+      .then(todos => {
+        return this.setState({
+          todos
+        });
+      })
+      .catch(err => console.error(err));
+    /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+  }
+
+  addTodo(newTodo) {
+    fetch('/api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    })
+      .then(res => {
+        if (!res.ok) {
+          throw new Error('Something went wrong.');
+        } else {
+          return res.json();
+        }
+      }
+      )
+      .then(todo => {
+        return this.setState({
+          todos: [...this.state.todos, todo]
+        });
+      })
+      .catch(err => console.error(err));
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    * add the created todo to the state array.
+    *
+    * Do not mutate the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+  }
+
+  toggleCompleted(todoId) {
+    const targetTodo = this.state.todos[(todoId - 1)];
```

Comment:
While this kind of works you can't rely on the `id` always being one off from the `index`. If the app supported delete the ids would get off from the index real quick. You want to use the `todoId` to find the correct index of the item. You can use something like `.findIndex`